app/controllers/api/v1/users_controller.rb

```diff
@@ -1,14 +1,36 @@
 # frozen_string_literal: true
 
+require 'securerandom'
 module Api
   module V1
     class UsersController < AuthenticatedController
-      before_action :require_authentication, except: [:recover_password, :create]
-      before_action :set_user, only: [:recover_password, :show, :update, :destroy]
+      before_action :require_authentication, except: [:recover_password, :recovery_token, :create]
+      before_action :set_user, only: [:show, :update, :destroy]
       before_action :require_authorisation, only: [:update, :destroy]
+      before_action :check_token, only: [:recover_password]
 
       def recover_password
-        PasswordRecoveryMailer.with(user: @user).recover_password.deliver_later
+        @user.password = SecureRandom.base64(10)
+        @user.token = nil
+
+        if @user.save
+          PasswordRecoveryMailer.with(user: @user).recover_password.deliver_later
+          head :ok
+        else
+          render 'api/v1/model_errors', locals: { errors: @user.errors }, status: :unprocessable_entity
+        end
+      end
+
+      def recovery_token
+        return head :bad_request unless params.key?(:email)
+
+        user = User.find_by(email: params[:email])
+
+        return head :not_found if user.nil?
+
+        token = SecureRandom.hex(16)
+        user.update!(token: token)
```

Comment:
How does the user know what's the generated token's value? Shouldn't an email be sent? 😄